### PR TITLE
Functional test for extension `oneapi::device_global`. Unique for every device

### PR DESCRIPTION
**Pull request by AndreyXRomanov**
This PR provides a functional test for `device_global`: The test checks that the value of the `device_global` instance will be unique for every device.

[Link](https://github.com/KhronosGroup/SYCL-CTS/pull/261/files/2230aa3012f49ea5f21e1f24164e136818868369..HEAD) to changes related to this PR

### CODE REVIEW

File: tests/extension/oneapi_device_global/device_global_common.h [hunk 1]
```diff
@@ -90,6 +90,17 @@ struct value_helper {
    */
   static void change_val(T& value, const int new_val = 1) { value = new_val; }
 
+  /**
+   * @brief The function changes value from the first parameter to
+   * value from the second parameter of the same type
```
Comment (keryell) [inline tests/extension/oneapi_device_global/device_global_common.h:95]:
```suggestion
   * value from the second parameter of the same type.
```

File: tests/extension/oneapi_device_global/device_global_common.h [hunk 2]
```diff
@@ -90,6 +90,17 @@ struct value_helper {
...
+   * Disabled if T is int to avoid function ambiguous
+   */
+  template <typename Ty = T>
+  static typename std::enable_if_t<!std::is_same_v<Ty, int>> change_val(
+      T& value, const T new_val) {
```
Comment (keryell) [inline tests/extension/oneapi_device_global/device_global_common.h:100]:
```suggestion
      T& value, const T& new_val) {
```
or remove the top-level `const`, as this is not fundamental here.

File: tests/extension/oneapi_device_global/device_global_common.h [hunk 3]
```diff
@@ -117,6 +128,17 @@ struct value_helper<T[N]> {
       value[i] = new_val;
     }
   }
+  /**
+   * @brief The function changes all values of the array from the first parameter to
+   * values of the array from the second parameter
+   * @param value The reference to the array that needs to be change
```
Comment (keryell) [inline tests/extension/oneapi_device_global/device_global_common.h:134]:
```suggestion
   * @param value The reference to the array that needs to be changed
```

File: tests/extension/oneapi_device_global/device_global_common.h [hunk 4]
```diff
@@ -142,6 +164,25 @@ struct value_helper<T[N]> {
...
+/** @brief The helper function to get variable address for pointer
+ *  @tparam T Type of variable
+ *  @param data variable to get pointer to
+ */
+template <typename T>
+inline T* pointer_helper(T& data) {
+  return &data;
+}
+
+/** @brief The helper function to get first element od array address for pointer
```
Comment (keryell) [inline tests/extension/oneapi_device_global/device_global_common.h:177]:
```suggestion
/** @brief The helper function to get first element of array address for pointer
```

File: tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        // Create sycl::queue instances for the subdevices and put it to the end
+        // of queues vector
+        queues.insert(queues.end(), subdevices.begin(), subdevices.end());
+      } else {
+        // If device doesn't support partition then create sycl::queue
+        // instance for the root device and put it to the end of queues vector
+        queues.emplace_back(device);
+      }
+    }
+    // If already have more that one device, then break the loop
```
Comment (keryell) [inline tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp:145]:
```suggestion
    // If already have more than one device, then break the loop
```

File: tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+      auto is_default_val_acc =
+          is_default_val_buf.template get_access<sycl::access_mode::write>(cgh);
+      cgh.single_task<kernel>([=] {
+        T def_value{};
+        is_default_val_acc[0] =
+            value_helper<T>::compare_val(dev_global<T>, def_value);
+      });
+    });
+  }
+  // Test fail if non-default value read from the device_global instance
```
Comment (keryell) [inline tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp:113]:
```suggestion
  // Test fails if non-default value read from the device_global instance
```

File: tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+  if (!is_default_val)
+    FAIL(log, get_case_description("device_global: Unique for every device",
+                                   "Value changed on another device. "
+                                   "Expect change only on one device",
+                                   type_name));
+}
+/**
+ * @brief The function tests that device_global instance unique for every device
```
Comment (keryell) [inline tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp:121]:
```suggestion
 * @brief The function tests that the device_global instance is unique for every device
```

**Review by keryell — CHANGES_REQUESTED** [review]

File: tests/extension/oneapi_device_global/device_global_common.h (hunk 1 above)
Comment (AndreyXRomanov) [reply tests/extension/oneapi_device_global/device_global_common.h]: Fixed

File: tests/extension/oneapi_device_global/device_global_common.h (hunk 2 above)
Comment (AndreyXRomanov) [reply tests/extension/oneapi_device_global/device_global_common.h]: Added reference to `new_val`

File: tests/extension/oneapi_device_global/device_global_common.h (hunk 3 above)
Comment (AndreyXRomanov) [reply tests/extension/oneapi_device_global/device_global_common.h]: Fixed typo

File: tests/extension/oneapi_device_global/device_global_common.h (hunk 4 above)
Comment (AndreyXRomanov) [reply tests/extension/oneapi_device_global/device_global_common.h]: Fixed

File: tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp (hunk 6 above)
Comment (AndreyXRomanov) [reply tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp]: Fixed

File: tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp (hunk 7 above)
Comment (AndreyXRomanov) [reply tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp]: Done

File: tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp (hunk 5 above)
Comment (AndreyXRomanov) [reply tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp]: Fixed

**Review by ProGTX — APPROVED** [review]

File: tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+              partition_property::partition_by_affinity_domain>(domain);
+          // Have to partition in at least 2 subdevices
+          if (sub_devices.size() > 1) {
+            return sub_devices;
+          }
+        }
+      }
+      // Return an empty vector if device doesn't support partition
+      case partition_property::no_partition:
+        return std::vector<sycl::device>();
```
Comment (keryell) [inline tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp:60]:
```suggestion
        return {};
```

File: tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+    }
+  }
+  // Return an empty vector if fails to partition in at least 2 devices
+  return std::vector<sycl::device>();
```
Comment (keryell) [inline tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp:64]:
```suggestion
  return {};
```

**Review by keryell — CHANGES_REQUESTED** [review]
Almost there!

File: tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp (hunk 9 above)
Comment (AndreyXRomanov) [reply tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp]: Done. Thank you

File: tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp (hunk 8 above)
Comment (AndreyXRomanov) [reply tests/extension/oneapi_device_global/device_global_functional_unique_for_every_device.cpp]: Done.

**Review by keryell — APPROVED** [review]
Thanks.

**Pull request merged**